content/blog/2023/sqlalchemy-for-django-developers.md

```diff
@@ -0,0 +1,351 @@
+2. SQLAlchemy for Django Developers (this article)
+3. Testing a FastAPI Application (coming soon)
+4. How To Use FastAPI Dependency Injection Everywhere (coming soon)
+
+[FastAPI Path Operations for Django Developers]: /2023/10/19/fastapi-path-operations-for-django-developers/
+
+## About SQLAlchemy
+
+Both Django's [ORM] (Object Relational Mapper) and [SQLAlchemy] are libraries
+for interacting with relational databases. Django's ORM is tightly coupled with
+the Django framework, while SQLAlchemy is a standalone library that can be used
+with any Python application. I heard about it many years ago, but never had the
+need to use it because Django's ORM was good enough for me (to the point that I
+know surprisingly little SQL). But when I started using FastAPI, SQLAlchemy was
+the recommended library for interacting with the database. Let's follow
+SQLAlchemy's [Quick Start] guide to create a simple database while learning
+about the differences between SQLAlchemy and Django's ORM.
+
+[ORM]: https://docs.djangoproject.com/en/4.2/topics/db/queries/
+[SQLAlchemy]: https://www.sqlalchemy.org/
+[Quick Start]: https://docs.sqlalchemy.org/en/20/orm/quickstart.html
+
+## Model Definition: Surprisingly Similar
+
+The first thing I noticed when I started using SQLAlchemy was how similar it is
+to Django's ORM when it comes to defining models as classes that represent
+database tables. Here is an example of a `User` model in Django:
+
+```python
+from django.db import models
+
+class User(models.Model):
+    name = models.CharField(max_length=100)
+    fullname = models.TextField()
+```
+
+And here is the same model in SQLAlchemy:
+
+```python
+from sqlalchemy import String
+from sqlalchemy.orm import Mapped, mapped_column, DeclarativeBase
+
+class User(DeclarativeBase):
+     __tablename__ = "user_account"
+
+     id: Mapped[int] = mapped_column(primary_key=True)
+     name: Mapped[str] = mapped_column(String(30))
+     fullname: Mapped[Optional[str]]
+```
+
+*Note: this [declarative style] for model definition is relatively new,
+superseding the old `declarative_base` function in SQLAlchemy 2.0. You might
+still encounter the old style in some codebases.*
+
+[declarative style]: https://docs.sqlalchemy.org/en/20/orm/mapping_styles.html#orm-declarative-mapping
+
+These two models are similar in that once they get information out of the
+database, table columns are accessible as attributes of the model instances such
+as `user.name` and `user.fullname`. In the same way, modifying the attributes of
+a model instance like `user.name = "Jane"` will modify the corresponding
+database record when the object is "saved" (which is done differently in
+SQLAlchemy as we will see later).
+
+In contrast, these models differ in that SQLAlchemy relies more heavily on type
+hints to define the model. In my opinion this results in a more robust developer
+experience, and the `Mapped` annotation enables some nice tricks we will see
```

Comment:
Would it be worth calling out later on what Mapped is enabling? I saw it was in use, but didn't see easily what the nice tricks were.

Reply:
Thanks, I tried to be more explicit in 835a72433fa9486a1d16f1a870ee0c1aa1e50a2c